EIPS/eip-2390.md

```diff
@@ -0,0 +1,299 @@
+
+GeoENS can do more than DNS-based solutions.
+In addition to geographic split horizon DNS, GeoENS can be used for the following:
+ - Locating digital resources (like smart contracts) that represent physical objects in the real world.
+ - Smart contract managing access to a physical object associated with a specific location.
+ - ENS + IPFS web hosting (as described in EIP1162) with content translated to the native language of the query source.
+ - Tokenizing objects with a physical location.
+
+Because of the decentralized nature of ENS, geo-specific resolution is different than traditional GeoDNS.
+GeoDNS works as follows. DNS queries are identified by their source IP address.
+This IP is looked up in a database like [GeoIP2](https://www.maxmind.com/en/geoip2-services-and-databases)
+from MaxMind which maps the IP address to a location.
+This method of locating the source of a query is error prone and unreliable.
+If the GeoIP database is out of date, queried locations can be vastly different than their true location.
+GeoENS does not rely on a database because the user includes a location in their query.
+
+It follows that queries can be made by users for any location, not just their location.
+Traditional DNS will only return the resource assigned to a query's provenance.
+GeoENS does not correlate a query's provinance with a location, allowing the
+entire globe to be queried from a single location.
+
+An additional shortcoming of traditional DNS is the fact that there is no way to return a list of servers in a certain proximity.
+This is paramount for uses cases that require discovering the resource with the lowest latency.
+GeoENS allows a list of resources, like IP addresses, to be gathered within a specific location.
+Then a client to determine themselves which resource has the lowest latency.
+
+Lastly, publicly facing GeoDNS services do not give fine granularity control
+over geographic regions for GeoDNS queries.
+Cloud based DNS services like [Amazon's Route 53](https://aws.amazon.com/route53/)
+only allow specifying geographic regions at the granularity of a State in
+the United States.
+GeoENS on the other hand gives 8 characters of geohash resolution which
+corresponds to +-20 meter accuracy.
+
+## Specification
+This EIP proposes a new interface to ENS resolvers such that geo-spacial information
+can be recorded and retrieved from the blockchain.
+The interface changes are described below for "address resolvers" described in EIP137
+however the idea applies to any record described in EIP1185 and EIP1062, namely DNS
+Resolvers, Text Resolvers, ABI Resolvers, etc.
+
+### What is a geohash?
+Geohashes are typically encoded in base 32 characters, as described
+https://en.m.wikipedia.org/wiki/Geohash#Algorithm_and_example.
+A geohash is an interleaving of latitude and longitude bits, whose
+length determines it's precision.
+
+
+### function setGeoAddr(bytes32 node, string calldata geohash, address addr) external authorised(node)
```

Comment:
This is not an editing comment, but a suggestion: strings are extremely expensive to deal with on the EVM. Based on "Geohashes are typically encoded in base 32 characters" and "GeoENS on the other hand gives 8 characters of geohash resolution" it seems they can easily fit in their unencoded (not base-32) form in a value type, such as uint256 or bytes32, and possible much smaller ones.

Reply:
Great idea - doing so reduced gas usage from 750498 to 749291 on average. New reference implementation provided here in commit 382c90848e402709dda8e60bbeb79d589155c7eb and the associated changes in the ENS repo https://github.com/james-choncholas/resolvers/commit/ba68579513eef9e71037e071b1097bdd7c9ec9b9